src/ronin/basefee-treasury/RoninBaseFeeTreasury.sol

```diff
@@ -0,0 +1,399 @@
+    _setContract(ContractType.STAKING, staking);
+    _setContract(ContractType.VALIDATOR, validatorSet);
+  }
+
+  function getThreshold() external view returns (Threshold memory threshold) {
+    return _threshold;
+  }
+
+  function setThreshold(uint8 num, uint8 denom) external onlyAdmin syncPendingProposals {
+    _setThreshold(num, denom);
+  }
+
+  function incrementNonce() external onlyAdmin syncPendingProposals {
+    _incrementNonce();
+  }
+
+  function voteFor(
+    address cid,
+    bytes32 hash,
+    Vote vote
+  ) external onlyCandidateAdmin(cid) onlyState(hash, State.Active) syncPendingProposals {
+    _vote(hash, cid, vote);
+    _tryExecute(hash);
+  }
+
+  function _vote(bytes32 hash, address cid, Vote vote) internal {
+    ProposalInfo storage $ = _info[hash];
+    IProfile profile = IProfile(getContract(ContractType.PROFILE));
+
+    _requireValidNonce($._proposal.nonce);
+    require(vote == Vote.For || vote == Vote.Against, ErrInvalidVote());
+    require($._votePower[cid] != 0, ErrNewlyRegisteredOrRenouncedOrEmergencyExit());
+    require($._vote[cid] == Vote.Unknown, ErrAlreadyVoted(hash, cid, $._vote[cid]));
+    require(
+      _computePeriod(profile.getId2RegisteredAt(cid) + COOLDOWN_PERIOD) <= _computePeriod($._createdAt),
+      ErrNewlyRegisteredCannotVote(cid)
+    );
+
+    $._vote[cid] = vote;
+    if (vote == Vote.For) {
+      $._currVotePower += $._votePower[cid];
+    }
+
+    emit Voted(msg.sender, cid, hash, vote, $._votePower[cid]);
+  }
+
+  function propose(
+    Proposal calldata proposal
+  ) external onlyCandidateAdmin(proposal.proposer) syncPendingProposals returns (bytes32 hash) {
+    validateProposal(proposal);
+
+    hash = hashProposal(proposal);
+
+    _requireState(hash, State.Unknown);
+
+    _createProposal(hash, proposal, _snapshotVotePower(hash));
+    _vote(hash, proposal.proposer, Vote.For);
+    _tryExecute(hash);
+  }
+
+  function _createProposal(bytes32 hash, Proposal memory p, uint256 totalVotePower) internal {
+    ProposalInfo storage $ = _info[hash];
+
+    $._createdAt = uint40(block.timestamp);
+    $._state = State.Active;
+    $._proposal = p;
+    $._minVotePower = _calcMinVotePower(totalVotePower);
+
+    require(_pendingProposals.add(hash), ErrExistedProposal(hash));
+
+    emit ProposalCreated(msg.sender, p.proposer, hash, p, $._createdAt);
+  }
+
+  function _snapshotVotePower(
+    bytes32 hash
+  ) internal returns (uint256 totalVotePower) {
+    address[] memory allCids = ICandidateManager(getContract(ContractType.VALIDATOR)).getValidatorCandidateIds();
+    uint256[] memory selfStakes = IStaking(getContract(ContractType.STAKING)).getManySelfStakingsById(allCids);
+
+    uint256 length = allCids.length;
+    require(length == selfStakes.length, ErrLengthMismatch(msg.sig));
+
+    ProposalInfo storage $ = _info[hash];
+
+    for (uint256 i; i < length; ++i) {
+      uint96 scaledStake = (selfStakes[i] / 1 ether).toUint96();
```

Comment:
The requirement doc doesn’t mention this downscaled logic

Reply:
I think it make sense since self stake for validator can never be < 1e18

Reply:
The edge case is that a validator with 10.01 ethers staked and another with 10.9 ethers staked have the same voting power. Does that make sense too?

Reply:
Why do we need to scale?

Reply:
@huyhuynh3103 as expected, make sense imo

Reply:
ok